File: characterReplacement.js
var characterReplacement = function (s, k) {
  let [longestSubstr, left, right, maxFreq] = [0, 0, 0, 0];
  const charsDict = {};

  while (right < s.length) {
    charsDict[s[right]] = (charsDict[s[right]] || 0) + 1;
    maxFreq = Math.max(maxFreq, charsDict[s[right]]);

    while (right - left + 1 - maxFreq > k) {
      charsDict[s[left]]--;
      left++;
    }

    longestSubstr = Math.max(longestSubstr, right - left + 1);
    right++;
  }
  return longestSubstr;
};

File: JavaWorkSpace/workspace/src/JavaPlayGround/CharacterReplacement.java
package JavaPlayGround;

public class CharacterReplacement {
    public int characterReplacement(String s, int k) {
        int answer = 0;
        int[] charCounts = new int[26];
        int maxFreq = 0;

        for (int right = 0, left = 0; right < s.length(); right++) {
            charCounts[s.charAt(right) - 'A']++;
            maxFreq = Math.max(maxFreq, charCounts[s.charAt(right) - 'A']);

            while (right - left + 1 - maxFreq > k) {
                charCounts[s.charAt(left) - 'A']--;
                left++;
            }

            answer = Math.max(answer, right - left + 1);
        }

        return answer;
    }
}
